yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,590 @@
+
+                self.nodes.push_back(temp)
+
+        for i in range(3):
+            self._left_edge[i] = self.nodes[0].left_edge[i]
+            self._right_edge[i] = self.nodes[0].right_edge[i]
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int split_helper(int start, int end, int num_children, int max_splits,
+                       np.int64_t * split_arr, np.float64_t * pos,
+                       np.int64_t * idx, np.float64_t * left,
+                       np.float64_t * right) nogil except -1:
+    cdef double lower = left[0]
+    cdef double upper = right[0]
+
+    split_arr[0] = start
+    split_arr[num_children] = end
+
+    split(0, num_children, max_splits, 0, 0, split_arr, idx, pos, left, right,
+          lower, upper)
+
+    return 0
+
+# This is a utility function which is able to catch errors when we attempt to
+# reserve too much memory
+cdef int reserve(vector[Node] * vec, int amount) except +MemoryError:
+    vec.reserve(amount)
+    return 0
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int seperate(np.float64_t * array, np.int64_t * idx, int offset,  double value,
+                  np.int64_t start, np.int64_t end) nogil except -1:
+    cdef np.int64_t index
+    cdef np.int64_t idx_index = start // 3, idx_split = idx_index
+    cdef np.int64_t split = start
+
+    for index in range(start, end, 3):
+        idx_index += 1
+        if array[index + offset] < value:
+            idx[idx_split], idx[idx_index] = idx[idx_index], idx[idx_split]
+            array[split], array[index] = array[index], array[split]
+            array[split+1], array[index+1] = array[index+1], array[split+1]
+            array[split+2], array[index+2] = array[index+2], array[split+2]
+            split+=3
+            idx_split+=1
+
+    return split
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int split(int start, int end, int max_splits, int splits, int axis,
+           np.int64_t * split_arr, np.int64_t * idx, np.float64_t * pos,
+           np.float64_t * left_edge, np.float64_t * right_edge,
+           np.float64_t lower, np.float64_t upper) nogil except -1:
+    cdef int mid
+    cdef double temp_value
+
+    if splits == max_splits and axis < 2:
+        splits = 0
+        axis += 1
+        lower = left_edge[axis]
+        upper = right_edge[axis]
+
+    if splits < max_splits:
+        splits += 1
+        mid = (start + end) // 2
+        temp_value = (upper + lower) / 2
+
+        split_arr[mid] = seperate(pos, idx, axis, temp_value, split_arr[start],
+                                  split_arr[end])
+
+        split(start, mid, max_splits, splits, axis, split_arr, idx, pos,
+              left_edge, right_edge, lower, temp_value)
+        split(mid, end, max_splits, splits, axis, split_arr, idx, pos,
+              left_edge, right_edge, temp_value, upper)
+
+    return 0
```

Comment:
this needs more comments, basically, this is capable of performing arbitary splits. When splitting into an oct you split the particle positons along the x axis into two groups, 
[small x] [large x]
you then split the two groups, by a y value
small x [small y] [large y], large x [small y] [large y]
and then again for z such that we get 8 children. This function is able to recursively do this to split into 8, 64, 512 etc